### ISSUE TITLE
package.json is included in the build files

### ISSUE THREAD

**Issue by piotrkmustardsystems** [theturtle32/WebSocket-Node#331]
Hi

I just noticed that your

lib\version.js file is looking for a version from package.json

`module.exports = require('../package.json').version;`

What this is doing internally, it exports all packege.json file as js module and adds it to the minified/uglified build.

So in my build, I got your whole package.json file with all dependencies and settings for no reason.

I also noticed that error with my build. I get around this by using node.env to pass it in like this:

version: process.env.VERSION

I do not know will this work for you but you will find a solution that fits your need.

Please modify this so your packege.json will not be added to any build that someone will make while using your library.

Regards

**Comment by ibc** [comment]
What is the problem exactly?

**Comment by ibc** [comment]
> Notice that the entire project.json is included in the compiled Javascript. Obviously it would be ideal if it wasn't.

Which is the problem about that? Obviously the package.json info is public info. It's common to expose the package version by requiring the `package.json` file (otherwise a separate `version` field is required somewhere else).

**Comment by francoabaroa** [comment]
@eloc3147 not sure if you're still having this issue.

I ran into this also, and had to add a clean script in my package.json scripts section

`
    "clean": "rm -rf node_modules/websocket/.git",`
`
    "heroku-postbuild": "npm run clean && npm install --only=dev --no-shrinkwrap",
`

Not sure what side effects it will have (none yet). My build is completing successfully now.

**Comment by socialcode-rob1** [comment]
.git needs to be removed from the npm published package

**Comment by piotrkmustardsystems** [comment]
@ibc Sorry but for my project packege.json is not a piece of public knowledge. We got our own git repos that I'm exposing to outside just because of this.

The way how you get the information from the package is including it and there is a simple solution to not do that.

1) you are making build bigger for the whole JSON object
2) you expose the packages and git repos to the outside world with is a security risk.

So this is needed to be fixed.

I do not get time to look at your class but if you point me where is your using the version I will show you how to change it to solve the problem.

Sorry for a long time not responding. I'm just following this but there is too much going on and just do not have time for this.

But I can help you with this one.

And yes I agree with @eloc3147, this is not only bad practice but a security risk in my project.

**Comment by ibc** [comment]
Guys, `.git` was included in `.npmignore` in websocket NPM 1.0.29 12 days ago. Have you verified it?

**Comment by piotrkmustardsystems** [comment]
Hi

I do not think that this will be a solution to the problem that I initially described.

But it's easy to test.

Import your package to anything.

Build it

Search in the build is your package included in js file.

If yes. Then this is not working :)

**Comment by ibc** [comment]
So you are talking about `package.json` being included and not about `.git` folder (which is already fixed in 1.29.0 but somehow commented again in this issue), right?

If so, you said:

> lib\version.js file is looking for a version from package.json
>
> module.exports = require('../package.json').version;
>
> What this is doing internally, it exports all packege.json file as js module and adds it to the minified/uglified build.
> So in my build, I got your whole package.json file with all dependencies and settings for no reason.

And what is the real problem with this?

**Comment by piotrkmustardsystems** [comment]
Ok, this is the line of code that is doing it.

I fixed this recently in one of my projects so I will give you later the solution.

Cause you required the file you are now pulling all packege.json file as an object into the production build.

Instead, just take the value from it and assigning just it to a variable.

The solution to this was to use process.env.VERSION

With I set up in env.js and assigned to process.env.VERSION.

When you 

require('./../package');

inside env.js file, this is not going to the build.

The wale is just set up and available to you with the node.js process rather than require that is pulling you object in.

And thanks to this packege.json is not part of your build.

And also not of a part of anybody who is using your package build as well. So for instance me :)

When I found this I also corrected all my code (cause I was previously doing the same thing) and it sorted this out.


So to summarize:

instead of:

module.exports = require('../package.json').version;

you should have your version pulled from env.js during the build or dev and set up like this:

module.exports = process.env.VERSION;

But you do not have env.js file (like I do in my react project) so you would set up this process variable in other more to fit your project structure.

**Comment by ibc** [comment]
I understand what you mean. Yes, indeed it would be better as you say. However I do not think that loading `package.json` into the build generates any real problem, does it? NOTE: I mean in this Node-WebSocket project which is public.

Said that, if you create a PR changing to your better approach we'll accept it :)

**Comment by piotrkmustardsystems** [comment]
Jest it better causes then I do not have your packege.json messing in my code.

I do not know your code. And how I see it you got a lot of people using this. I do not have time to correct cause I'm not familiar with your code. It's your project.

I already did you a favor and pointed this out. Correction should take you max 1h of work. And I pointed this out 8 months ago.

Now you get what the problem is. What the solution will be. You know the code and how to set it up.

Sorry, I can't do this for you. But you should definitely do it.

**Comment by piotrkmustardsystems** [comment]
I think if you are using npm start you can use it like this and the problem should be gone:

https://stackoverflow.com/a/22339262

The top solution in this thread is also suggesting as you do but I think somebody did not think this through. But I also did not know this 8 months a go.

**Comment by ibc** [comment]
> Now you get what the problem is. What the solution will be. You know the code and how to set it up.
> Sorry, I can't do this for you. 

OK, I'll do it when it really fixes a real issue and/or I get lot of free spare time for cosmetic changes.

> But you should definitely do it.

Believe me: No, I don't.

**Comment by piotrkmustardsystems** [comment]
This is not a cosmetic change.

I see you're not getting the problem with it. Some people really care about code size and security issues.

Cosmetic is the solution. But I do not know how to apply it to your code, cause I do not know it as you do.

So I do not understand why did you close this issue ?? It's not done and it's not cosmetic.
